source/adios2/helper/adiosXML.cpp

```diff
@@ -11,473 +11,162 @@
-                " can't have attributes, in call to ADIOS constructor\n");
-        }
-    }
-    else // opening tag
-    {
-        attributes = lf_GetAttributes(openingTag);
-    }
-    return attributes;
-}
-
-void InitXML(const std::string configXML, const MPI_Comm mpiComm,
-             const bool debugMode,
-             std::vector<std::shared_ptr<Transform>> &transforms,
-             std::map<std::string, IO> &ios)
-{
-    // independent IO
-    std::string fileContents(FileToString(configXML));
-    if (fileContents.empty())
-    {
-        // issue a warning?
-        return;
-    }
-
-    RemoveCommentsXML(fileContents);
-
-    // adios-config
-    std::string::size_type position(0);
-    const TagXML adiosConfigXML(
-        GetTagXML("adios-config", fileContents, position));
-
-    // process transforms, not yet implemented
-
-    while (position != std::string::npos)
-    {
-        const TagXML transformXML(
-            GetTagXML("transform ", adiosConfigXML.Elements, position));
-
-        if (transformXML.Header.empty())
-        {
-            break;
+            continue;
         }
-        // InitTransform(transformTag, debugMode, transforms);
-    }
-
-    position = 0;
-    // process IOs
-    while (position != std::string::npos)
-    {
-        // io
-        const TagXML ioXML(GetTagXML("io ", adiosConfigXML.Elements, position));
 
-        if (ioXML.Header.empty()) // no more groups to find
-        {
-            break;
-        }
-        InitIOXML(ioXML, mpiComm, debugMode, transforms, ios);
+        params.emplace(attrKey.value(), attrValue.value());
     }
+    return params;
 }
 
-void InitIOXML(const TagXML &ioXML, const MPI_Comm mpiComm,
+void InitIOXML(pugi::xml_node ioNode, const MPI_Comm mpiComm,
                const bool debugMode,
                std::vector<std::shared_ptr<Transform>> &transforms,
                std::map<std::string, IO> &ios)
 {
-    const Params ioAttributes(GetTagAttributesXML(ioXML.Header));
-
-    std::string ioName;
-    for (const auto &ioAttribute : ioAttributes)
+    // Extract <io name=""> attribute
+    pugi::xml_attribute nameAttr = ioNode.attribute("name");
+    if (!nameAttr)
     {
-        if (ioAttribute.first == "name")
-        {
-            ioName = ioAttribute.second;
-        }
-    }
-
-    if (debugMode)
-    {
-        if (ioName.empty())
-        {
-            throw std::invalid_argument(
-                "ERROR: io name=\"value\" attribute not found in opening XML "
-                "tag " +
-                ioXML.Header +
-                ", check XML config file, in call to ADIOS constructor\n");
-        }
-
-        if (ios.count(ioName) == 1) // io exists
+        if (debugMode)
         {
-            throw std::invalid_argument("ERROR: io name " + ioName +
-                                        " must be unique in XML config file, "
-                                        "in call to ADIOS constructor\n");
```

Comment:
Let's be consistent in the exception error messages. These are for the user to report bugs. The more informative the better, use std::invalid_argument to be more specific and that this is coming from the XML config file and the ADIOS constructor. Keep in mind that adios is typically just one component of many. It's nice to know where the ERROR comes from in the user code.